### DISCUSSION TITLE
Trainer won't achieve Data Parallelism

### DISCUSSION THREAD

**Discussion by b1nch3f** [coqui-ai/Trainer#54] [Q&A]
Data Parallelism is when we split the mini-batch of samples into multiple smaller mini-batches and run the computation for each of the smaller mini-batches in parallel. <sup>[1]</sup>

Correct me if i'm wrong, but the following code creates subprocess, so basically the same script will take the entire mini-batch. In a multiple GPU situation, it doesn't look the code will exploit multiple GPUs because it's not written to achieve data-parallelism. 
https://github.com/coqui-ai/Trainer/blob/0b3c88a2d8185dd31cc8d9a2d60d80a48e08da33/trainer/distribute.py#L54

Additionally, let's say I put save checkpoint at 100th iteration, not sure if the trainer is capable to save that checkpoint with the least loss, it's like at 100th iteration, the script that finishes last will overwrite the checkpoints saved by the scripts ran previously on other GPUs.

I looked at trainer code after fine-tuning on 4 GPUs, initially i ran it on 2 GPUs and expected a quicker run with 4 GPUs, but that didn't happen and it had the same runtime like 2 GPUs.

Ref.
1. https://pytorch.org/tutorials/beginner/former_torchies/parallelism_tutorial.html

**Comment by erogol** [comment]
You need to set your data loader accordingly if you intend to use DataParallelism. 👟 only wraps your model for DDP and runs it in the right way for DDP. The rest relies on your model implementation. So a basic change would be using DistributedSampler to send different splits of data to different GPUs.
